### DISCUSSION TITLE
Lambda route example?

### DISCUSSION THREAD

**Discussion by seikosantana** [scottoffen/grapevine#74] [Q&A] [answered]
I'm trying to create routes with lambda, and here's what i've got.

```csharp
static void Main(string[] args)
{
    Console.WriteLine("Hello World!");
    var testRoute = new Route(async (context) =>
    {
        await context.Response.SendResponseAsync("Test");
    }, "GET", "/text");

    var server = RestServerBuilder.UseDefaults().Build();
    server.Start();
    Console.ReadLine();
}
```

The route scanning is completed and found 0 routes.

I'm pretty sure i am missing something, but i can't find any documentations or examples about creating and also using the route on the server. What is the correct way to use the lambda route?

Edit: I tried doing this and it returns not implemented.

```csharp
static void Main(string[] args)
{
    Console.WriteLine("Hello World!");
    var testRoute = new Route(async (context) =>
    {
        await context.Response.SendResponseAsync("Test");
    }, "GET", "/text");

    var server = RestServerBuilder.UseDefaults().Build();
    server.UseCorsPolicy();
    server.Router.Register(testRoute);
    server.Start();
    Console.ReadLine();
}
```

Thanks in advance

**Comment by seikosantana** [comment, marked as answer]
Nevermind. Above code works, wonder why it didn't before. But I saw in the samples that the route is registered on AfterStarting event. Does it make any difference? Is it fine if i just register routes before starting?

**Reply by scottoffen** [reply]
During server start, the server checks to see if any routes have already been registered. If not, it auto scans for routes ([see code here](https://github.com/scottoffen/grapevine/blob/main/src/Grapevine/RestServer.cs#L145)). But if you have already registered routes, it assumes that you don't want to auto scan.

If you add a lambda route **before the server starts**, auto scanning will not happen.

If you want auto scanning to happen AND you want to add lambda routes, the recommended way to do this is to add them after the server has done the auto scanning. The easiest way to do this is by adding a handler to `Server.AfterStarting`.

If you want the lambdas first AND you want to auto scanning, you will need to do the auto scanning in a `Server.AfterStarting` handler.

```csharp
server.AfterStarting += (s) => s.Router.Register(RouteScanner.Scan());
```

But I think I might change that behavior by adding an options to force auto scanning even if routes are already registered. It would simplify this for people. What do you think?

**Reply by seikosantana** [reply]
Thanks

In my opinion, the current behavior is better. I would prefer registering routes manually since i won't have to create a separate class for the `RestResource`. What I have in mind is similar to what Express JS have, so maybe i can create a route and then expose a route for the server to use or something.

It's simple enough to me whether to scan or not scan routes with the example you specified above, maybe it just needs to be shown in documentation or something.
